src/librustdoc/visit_ast.rs

```diff
@@ -337,4 +352,58 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
             om.foreigns.push((item, renamed));
         }
     }
+
+    /// Run a few checks on the `#[doc]` attributes for an item:
+    /// - warn if a crate-level attribute is used on a non-`crate` item
+    /// - warn if `#[doc(inline)]`/`#[doc(no_inline)]` is used on a non-`use` item
+    /// - error if conflicting `#[doc(inline)]`/`#[doc(no_inline)]` are found
+    fn check_item_attrs(&mut self, attrs: &'tcx [ast::Attribute], is_crate: bool, is_use: bool) {
```

Comment:
nit: this allows impossible states (`is_crate` and `is_use` set simultaneously). I'd prefer to have an explicit type instead:

```rust
enum AttrKind {
    Crate,
    Use,
    Other,
}
```
but at that point it might be simpler to just pass in the `ItemKind` directly.

Reply:
There is no `ItemKind::Crate`, so I'll go with an `AttrKind` enum